js/pathfinding.js

```diff
@@ -5,56 +5,250 @@ import {settingsKey} from "./settings.js";
+	}
+
+	ensureCacheExists(cacheId) {
 		// Track that we've last used this cache right now
 		this.lastUsed.set(cacheId, Date.now());
 
 		// Get the nodes for the cacheId. If we don't already have one, create one
-		let cachedNodes = this.nodes.get(cacheId);
-		if (!cachedNodes) {
-			cachedNodes = new Array(gridHeight);
+		if (!this.nodes.has(cacheId)) {
+			const cachedNodes = new Array(gridHeight);
 			for (let y = 0; y < gridHeight; y++) {
 				cachedNodes[y] = new Array(gridWidth);
 				for (let x = 0; x < gridWidth; x++) {
 					cachedNodes[y][x] = {x, y};
 				}
 			}
 			this.nodes.set(cacheId, cachedNodes);
+			this.background.queues.set(cacheId, new ProcessOnceQueue());
 
 			// Since we're adding a new cache, check if we have too many and,
 			// if we do, get rid of the one that was last used longest ago
 			if (this.lastUsed.size > Cache.maxCacheIds) {
-				let oldest;
-				for (let entry of this.lastUsed) {
-					if (!oldest || oldest[1] > entry[1]) {
-						oldest = entry;
-					}
-				}
-				this.nodes.delete(oldest[0]);
-				this.lastUsed.delete(oldest[0]);
+				// Build an array from the last used entries, sort in ascending order, and get the first
+				// element (with the lowest last-used value) as the oldest entry, then retrieve the cache ID
+				// from that element
+				const oldestCacheId = Array.from(this.lastUsed.entries()).sort((a, b) => a[1] - b[1])[0][0];
+
+				this.nodes.delete(oldestCacheId);
+				this.lastUsed.delete(oldestCacheId);
+				this.background.queues.delete(oldestCacheId);
+			}
+		}
+	}
+
+	/**
+	 * Get the cache associated with the given cache ID, creating a new one
+	 * if we don't already have one
+	 */
+	getCachedNodes(token) {
+		const cacheId = this.getCacheId(token);
+		return this.nodes.get(cacheId);
+	}
+
+	/**
+	 * Start background caching from the token's current position
+	 */
+	startBackgroundCaching(token) {
+		const cacheId = this.getCacheId(token);
+		this.background.queues.get(cacheId).push(
+			{
+				value: {
+					pos: getGridPositionFromPixelsObj(token.position)
+				},
+				cacheId,
+				token
 			}
+		);
+		this.scheduleBackgroundCache();
+	}
+
+	/**
+	 * Find if any of the caches have more nodes to background cache. If there is, then schedule a background
+	 * caching job for that queue
+	 */
+	scheduleBackgroundCache() {
+		// If we already have a nextJobId, then don't start another one
+		if (this.background.nextJobId) return;
+
+		// Find the latest-used cache that has nodes left to cache
+		const latestCacheId = this.getLatestCacheId();
+		if (latestCacheId) {
+			this.background.nextJobId = window.requestIdleCallback(
+				() => this.runBackgroundCache(this.background.queues.get(latestCacheId))
+			);
+			this.scheduleTimeout();
+		}
+	}
+
+	/**
+	 * Start a timeout which, if we read the timeout time, will schedule a small amount of caching
+	 * to be performed every frame. This timeout will be cancelled every time we perform background caching.
+	 */
+	scheduleTimeout() {
```

Comment:
Maybe rename this to `kickWatchdog` to make it more clear what the timeout's purpose is.